### DISCUSSION TITLE
What is the v5 equivalent of `workflow.add_task()`?

### DISCUSSION THREAD

**Discussion by dmerrick** [argoproj-labs/hera#605] [Q&A]
Related to https://github.com/argoproj-labs/hera/issues/579, now that I have my Script created I'd like to attach it to a Workflow. We used to use `workflow.add_task()`, but that seems to have been removed in v5.

Do we have to migrate to the contextmanager pattern, or is there a Hera v5 equivalent to `add_task()`?

**Comment by elliotgunton** [comment]
Hi @dmerrick, with v5 we split out the coupling of tasks/templates, so the equivalent of `workflow.add_task` requires the `Workflow` and a `DAG` under it. Then, adding a `Task` under the context manager is the recommended/only way as we haven't added a public API for something like `dag.add_task()`. If you have your script using the `script` decorator on a function, then you can follow the main README example:

```py
from hera.workflows import DAG, Workflow, script

@script()
def echo(message: str):
    print(message)

with Workflow(
    generate_name="dag-diamond-",
    entrypoint="diamond",
) as w:
    with DAG(name="diamond"):
        A = echo(name="A", arguments={"message": "A"})
        B = echo(name="B", arguments={"message": "B"})
        C = echo(name="C", arguments={"message": "C"})
        D = echo(name="D", arguments={"message": "D"})
        A >> [B, C] >> D

w.create()
```


A more manual approach (but still using the context manager and implicitly adding your task) would be:
```py
from hera.workflows import DAG, Script, Workflow

with Workflow(
    generate_name="one-script-task-",
    entrypoint="one-script-task",
) as w:
    my_script = Script(name="my-script", ...)  # implicitly adds the "my-script" template to your workflow
    with DAG(name="one-script-task"):
        Task(name="A", template=my_script, arguments={...})  # implicitly adds the task "A" to the "one-script-task" dag

w.create()
```

If you'd prefer an explicit, public API way to do `add_task`, we can add a new issue/discussion in the Github (as it would involve a few things - adding APIs to workflows, DAGs and Steps) - I'll ping @flaviuvadan and @samj1912 here to weigh in but I think we want to keep to the one method of building Workflows with context managers.
